streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceivedBlockTracker.scala

```diff
@@ -235,8 +297,84 @@ private[streaming] class ReceivedBlockTracker(
     }
   }
 
+  /**
+   * Creates a WAL Writer in a separate thread to enable batching of log events.
+   * Exposed for tests.
+   */
+  protected def createBatchWriteAheadLogWriter(): Option[BatchLogWriter] = {
+    if (!WriteAheadLogUtils.isBatchingEnabled(conf)) return None
+    val writer = checkpointDirOption.map(_ => new BatchLogWriter)
+    writer.foreach { runnable =>
+      val thread = new Thread(runnable, "Batch WAL Writer")
+      thread.setDaemon(true)
+      thread.start()
+    }
+    writer
+  }
+
   /** Check if the write ahead log is enabled. This is only used for testing purposes. */
   private[streaming] def isWriteAheadLogEnabled: Boolean = writeAheadLogOption.nonEmpty
+
+  /** A helper class that writes LogEvents in a separate thread to allow for batching. */
+  private[streaming] class BatchLogWriter extends Runnable {
+
+    var active: Boolean = true
+
+    private def writeRecords(records: Seq[ReceivedBlockTrackerLogEvent]): Unit = {
+      writeAheadLogOption.foreach { logManager =>
+        if (records.nonEmpty) {
+          logDebug(s"Batched ${records.length} records for WAL write")
+          logManager.write(ByteBuffer.wrap(Utils.serialize(
+            CombinedReceivedBlockTrackerLogEvent(records))), clock.getTimeMillis())
+        }
+      }
+    }
+
+    def stop(): Unit = {
+      logInfo("Stopping Batch Write Ahead Log writer.")
+      active = false
+    }
+
+    private def flushRecords(): Unit = {
+      val buffer = new ArrayBuffer[ReceivedBlockTrackerLogEvent]()
+      try {
+        buffer.append(walWriteQueue.take())
+        val numBatched = walWriteQueue.drainTo(buffer.asJava) + 1
+        logDebug(s"Received $numBatched records from queue")
+      } catch {
+        case _: InterruptedException =>
+          logWarning("Batch Write Ahead Log Writer queue interrupted.")
+      }
+      def updateRecordStatus(record: ReceivedBlockTrackerLogEvent, successful: Boolean): Unit = {
+        val promise = walWriteStatusMap.get(record)
+        if (promise == null) {
+          logError(s"Promise for writing record $record not found in status map!")
+        } else {
+          promise.success(successful)
+        }
+      }
+      try {
+        writeRecords(buffer)
+        buffer.foreach(updateRecordStatus(_, successful = true))
+      } catch {
+        case NonFatal(e) =>
+          logWarning(s"Batch WAL Writer failed to write $buffer", e)
+          buffer.foreach(updateRecordStatus(_, successful = false))
+      }
+    }
+
+    override def run(): Unit = {
```

Comment:
nit: would be nice to order the public methods above the private methods. easier to read the control flow.